pkg/router/controller/route_secret_manager.go

```diff
@@ -118,106 +192,199 @@ func (p *RouteSecretManager) HandleRoute(eventType watch.EventType, route *route
 	// secret handler
 	return cache.ResourceEventHandlerFuncs{
 
-		// AddFunc is intentionally left empty (only logs the event) because this handler is generated only after ensuring the existence of the secret.
-		// By leaving this empty, we prevent unnecessary triggering for the addition of the secret again. Additionally, GetSecret() method is called
-		// immediately after registering with the secretManager, to read the secret from the cache.
 		AddFunc: func(obj interface{}) {
 			secret := obj.(*kapi.Secret)
-			log.V(4).Info("secret added for route", "namespace", route.Namespace, "secret", secret.Name, "route", route.Name)
+			log.V(4).Info("secret added for route", "namespace", namespace, "secret", secret.Name, "route", routeName)
+
+			// Secret re-creation scenario
+			// Check if the route key exists in the deletedSecrets map, indicating that the secret was previously deleted for this route.
+			// If it exists, it means the secret is being recreated. Remove the key from the map and proceed with handling the route.
+			// Otherwise, no-op (new secret creation scenario and no race condition with that flow)
+			// This helps to differentiate between a new secret creation and a re-creation of a previously deleted secret.
+			key := generateKey(namespace, routeName)
+			if _, deleted := p.deletedSecrets.LoadAndDelete(key); deleted {
+				msg := fmt.Sprintf("secret %q recreated for route %q", secret.Name, key)
+				log.V(4).Info(msg)
+
+				// Ensure fetching the updated route
+				route, err := p.getUpdatedRoute(namespace, routeName)
+				if err != nil {
+					return
+				}
+
+				// The route should *remain* rejected until it's re-evaluated
+				// by all the plugins (including this plugin). Once passes, the route will become active again.
+				p.recorder.RecordRouteRejection(route, "ExternalCertificateSecretRecreated", msg)
+			}
 		},
 
 		UpdateFunc: func(old interface{}, new interface{}) {
 			secretOld := old.(*kapi.Secret)
 			secretNew := new.(*kapi.Secret)
-			log.V(4).Info("secret updated for route", "namespace", route.Namespace, "secret", secretNew.Name, "old-version", secretOld.ResourceVersion, "new-version", secretNew.ResourceVersion, "route", route.Name)
-
-			// re-validate
-			fldPath := field.NewPath("spec").Child("tls").Child("externalCertificate")
-			if err := routeapihelpers.ValidateTLSExternalCertificate(route, fldPath, p.sarClient, p.secretsGetter).ToAggregate(); err != nil {
-				log.Error(err, "skipping route due to invalid externalCertificate configuration", "namespace", route.Namespace, "route", route.Name)
-				p.recorder.RecordRouteRejection(route, "ExternalCertificateValidationFailed", err.Error())
-				p.plugin.HandleRoute(watch.Deleted, route)
-				return
-			}
+			key := generateKey(namespace, routeName)
+			msg := fmt.Sprintf("secret %q updated for route %q (old resourceVersion=%v, new resourceVersion=%v)", secretNew.Name, key, secretOld.ResourceVersion, secretNew.ResourceVersion)
+			log.V(4).Info(msg)
```

Comment:
Same comment about the structured logging.

Reply:
Thanks, updated